### ISSUE TITLE
Inconsistent connection update behavior, potentially causing client_secret deletion

### ISSUE THREAD

**Issue by lundybernard** [auth0/auth0-deploy-cli#420]
### Description

deploying config `a0deploy deploy -c tenant/config.json --input_file tenant/tenant.yaml`

results in error:
> error: Problem running command deploy during stage processChanges when processing type connections
> error: Problem updating connections {"name":"###","id":"###"}
> Bad Request: Before creating Windows Azure Active Directory connections you must configure the client_id and client_secret settings in the dashboard. For more information go to https://docs.auth0.com/waad-clientid

also:
> error: Problem running command deploy during stage processChanges when processing type connections
> error: Problem updating connections {"name":"####","id":"####"}
> Bad Request: The signing certificate is not valid. Please provide a .pem or .cer certificate base64 encoded public key

### Reproduction
1. Export an existing environment, including one or more AzureAD connections

We are stripping out some sensitive information on export

config.json:
```
{
...
  "AUTH0_ALLOW_DELETE": false,
  "AUTH0_EXPORT_IDENTIFIERS": false,
  "AUTH0_EXCLUDED_RULES": [ "rule-1-name" ],
  "AUTH0_EXCLUDED_CONNECTIONS" : [
    "####",
    "smoke-test"
  ],
  "EXCLUDED_PROPS": {
    "connections": [
        "options.client_secret",
        "options.admin_access_token",
        "options.admin_access_token_expiresin",
        "options.cert",
        "options.signingCert"
    ]
  },
    "INCLUDED_PROPS": {
      "rulesConfigs": [ "key" ]
  }
}
```
tenant.yaml:
```
connections:
  - name: ####
    strategy: waad
    enabled_clients:
      - ####
    is_domain_connection: false
    options:
      domain: ####.com
      granted: true
      tenantId: ####
      client_id: ####
      use_wsfed: false
      app_domain: ####.auth0.com
      ext_groups: true
      ext_profile: true
      thumbprints:
        - ####
        - ####
        - ####
        - ####
        - ####
        - ####
      identity_api: azure-active-directory-v1.0
      basic_profile: false
      tenant_domain: ####.com
      waad_protocol: openid-connect
      domain_aliases:
        - ####.com
      api_enable_users: false
      ext_nested_groups: true
      useCommonEndpoint: false
      should_trust_email_verified_connection: ####
    show_as_button: false
```

Deploy the config: `a0deploy deploy -c tenant/config.json --input_file tenant/tenant.yaml`

the deployment halts with the given error.

The strange thing is that this worked perfectly in 2 other environments, with almost identical configurations (azuread and samlp connections) just a few days ago.  Re-testing those environments they now throw the same errors. It looks like these options are now required instead of optional.

#### Workaround:
for now, you can add the failing connections to the AUTH0_EXCLUDED_CONNECTIONS list to skip them entirely

### Environment

- **Version of this library used:**
latest = 7.3.7

**Comment by lundybernard** [comment]
We noticed further problems with this deployment, after excluding the connections which threw an error ( two saml and one azure connections), the deployment succeeded, but it also wiped out the client_secret, which caused failed connections.  Fortunately we were able to restore those secret keys from a backup, but this is new behavior which may cause serious issues for others.

**Comment by willvedd** [comment]
@lundybernard Thanks for raising this. There's a bit to digest here, can you confirm my understanding of this issue? You have some connection (Azure AD?) that you've explicitly excluded the `client_secret` property from through the `EXCLUDED_PROPS` configuration but when running an import, you're observing that the Deploy CLI has removed the `client_secret` (and possibly other) connection properties from your connections, causing them to fail. Is this correct?

And as a workaround, you've explicitly excluded those connections through the `AUTH0_EXCLUDED_CONNECTIONS` configuration?

**Comment by willvedd** [comment]
After looking into this further, I think I've nailed down what has happened. 

While attempting to create a new tenant from previous configurations, you incurred the two errors above. Those errors are on-point about the culprits, they lack the necessary configurations to create a new Azure AD connection. The reason why those configurations are lacking is because they've been explicitly excluded in the `config.json` file; confirmed by the lack of `client_secret ` and `cert` properties in your `tenant.yml` file. So far this makes sense. In order to create the Azure AD connection, you need to provide `client_id` and `client_secret`, as the error had indicated. I was able to look at the relevant API code and confirmed that is has not changed in many years, so that rules out the API regression.

Where I believe the confusion is occurring is around these errors being the cause of any destructive changes. The existence of these errors means that the Azure AD connection was not able to be created, or even updated if it had existed on the tenant beforehand. Excluding this connection from your workflow acts as a way to workaround those errors, but the issue persists. I do not have any reason to believe that the client_secret was wiped out because those errors lead me to believe that the connection never existed in the first place. 

As for the case where a connection currently exists and then the Deploy CLI is used to update that connection with `client_secret` removed, this will not work because of how the [update connection API](https://auth0.com/docs/api/management/v2#!/Connections/patch_connections_by_id) functions. Specifically this snippet (emphasis mine):

>Note: if you use the **options** parameter, the whole options object will be overridden, so ensure that all parameters are present

Meaning, when the `options` are patched, there is no reconciliation between what configuration exists currently for that connection and what you're proposing. The `options` object being sent in the PATCH payload will overwrite any existing configuration. This may in theory cause certain parameters to be wiped out, but I don't believe this is the case with the Azure AD connection because it validates the existence of `client_secret` before applying the updates.

Wether or not this is ideal behavior is up for debate. I do think the Deploy CLI can do a better job showcasing the proposed changes against what exists on your tenant, much like a `terraform plan` does. This has been an often asked feature.

**Comment by lundybernard** [comment]
to clarify, we have excluded the `client_secret` from all connections via
```
"EXCLUDED_PROPS": {
"connections": [
    "options.client_secret",
    ...
```

As I understand it now, this will only exclude those properties when exporting an existing configuration, and is not taken into account when deploying changes.

so, I have a tenants.yaml config that includes all of the connections without the client_secret attribute.

The first problem, is that running deploy failed because of the missing parameter on the azureAD connections.  the work-around is to exclude those connections  in `AUTH0_EXCLUDED_CONNECTIONS` and configure them manually.  Our prefered solution would be for the API to ignore excluded parameters.  if client_secret is not specified it should not be changed, if we want it removed we should specify an empty string or `None` value.

the second, more serious problem, involves a Google apps connection `strategy: google-apps`, the `client_secret` was also excluded from the exported yaml for this connection, but when we ran deploy it did not throw an error, instead it appears that its client_secret field was over-written, we assume blanked out.  We experienced login errors with that connection immediately after running deploy, and resolved them by plugging the client_secret back in through the admin interface, which fixed the problems.

As a workaround we are currently excluding ALL connections via the `AUTH0_EXCLUDED_CONNECTIONS`, which It appears only affects PUSHING configs to Auth0, as excluded connections still show up in tenants.yaml when exporting a config.  that makes this request (https://github.com/auth0/auth0-deploy-cli/issues/375) a higher priority for us.  Also, clarification on what parameters affect import vs export, or parity across those two functions, would be helpful.  I expect excluded connections to be excluded in both directions, and its surprising to see them included in an export but excluded during import.

**Comment by kanocarra** [comment]
I am also having this same issue, it always appears to wipe my Google social connection `client_secrent` and `client_id` on deployment, even with the specified:
```
  "EXCLUDED_PROPS": {
    "connections": ["options.client_secret","options.client_id"]
  }
```

You are right in that this only seems to work when importing an existing configuration from auth0 and not when it is deploying a new one. Interestingly though, the client secret for my Github social connection still appears to be there - is it just an issue with some social connections? 

I guess I will also have to just exclude the whole connection but not ideal!

**Comment by willvedd** [comment]
@lundybernard Appreciate the follow-up. I am able to confirm your observations, there is inconsistent behavior within the [PATCH connections endpoint](https://auth0.com/docs/api/management/v2#!/Connections/patch_connections_by_id) depending on the connection type. 

**Azure AD:** Validates payload for the existence of client_id and client_secret, returns error.
**Google Apps:** Does not validate payload for client_secret . Succeeds and deletes client_secret.

It appears that there are some reasons why these differences exist, but going to work further with the API team to see if we can bring more parity there. 

I'm also going to work with our team to implement warning and safeguards in this project to prevent `client_secret` and other important properties from being deleted to existing resources; I don't believe there is a strong-enough use case to allow this behavior. 

However, in order to provide a more long-term fix, can you provide an explanation for excluding those properties? I understand that they're sensitive values, are you looking to prevent them from being written to plain-text?

**Comment by kanocarra** [comment]
@willvedd 
> Can you explain what your use case is for excluding `client_id` and `client_secret` from those connections?

Our use case is the following:

* We are an AWS shop so all AWS things are automated, for Google social connection, we have a Google Cloud account that we generate the OAuth credentials screen, client ID and client secret from. Because we only need this on production, we don't bother automating the creation of the oauth keys and instead do it from the Google Cloud console once.
* We'd prefer to not have `client_id` and `client_secret` stored in any deployment configs or our CI server (we use CircleCI so env vars become difficult to manage). 
* Therefore, our use case is to run the deployment to auth0, specifying the google-oauth2 social connection _without_ `lcient_id` or `client_secret` but then being able to manually copy/paste the client ID and client secret into auth0 directly. This only has to be done once and we can easily rotate the secret when required without deployment.

The expected behaviour is for auth0 to ignore the `client_id` and `client_secret` when a deployment (`a0deploy export`) is run. Currently instead, it _overwrites_ the existing client ID and secret to blank. This is super dangerous because then auth0 defaults to the development keys in a production environment. We often are deploying changes to our auth0 config via the CLI so this is quite a concerning issue for us. 

I don't use the `a0deploy import` function as we already have our desired config defined in yaml, I would expect `EXCLUDED_PROPS` to work both ways but it appears to only work on the `import`?

**Comment by lundybernard** [comment]
@willvedd 

Our use case is much the same as kanocarra.  

- Essentially we are using a0deploy as a configuration management system.
  - We want to store the configuration in a git repo
    - so that changes are subject to review
    - history is recorded
  - We use the export (pull) functionality to diff the configs and look for changes made outside of change-control
  - We use the import (push) functionality:
    - to maintain a single source of truth
    - for disaster recovery
    - to coordinate changes during deployments
  - Long term we would like to setup a base template, and over-ride configs for each environment, which I believe is already possible.
- We do not want sensitive information stored in our github repositories.
  - Long term we may use environment variables to inject those secrets in a secure way.


I also expect EXCLUDED_PROPS and EXCLUDED_CONNECTIONS, and other EXCLUDE* settings to be bi-directional, essentially acting like entries in .gitignore.  Better documentation about what settings affect export/import would be valuable.

Thanks for your help!

**Comment by willvedd** [comment]
@lundybernard Understood, thanks for confirming. I think this is a reasonable use case and enough for us to explore some possible solutions in the coming weeks.

**Comment by kanocarra** [comment]
> I also expect EXCLUDED_PROPS and EXCLUDED_CONNECTIONS, and other EXCLUDE* settings to be bi-directional, essentially acting like entries in .gitignore. Better documentation about what settings affect export/import would be valuable.

This was also a big thing for me, I had assumed they were bi-directional and the documentation makes no mention that it will only be adhered to on `import`. 

In saying that, I think this tool is great and works for 99% of tasks I want it for. It's just unfortunate that this small issue has such massive impact.

Thanks for all your hard work!

**Comment by willvedd** [comment]
>This was also a big thing for me, I had assumed they were bi-directional and the documentation makes no mention that it will only be adhered to on import.

My understanding is that this only affects properties nested under the `options` property for connections. This is because most other endpoints adhere to normal PATCH conventions where excluded properties are ignored. However, as we can see in the endpoint docs, the `options` property does not adhere to that convention. So for most other circumstances it turns out to work bi-directionally.

**Comment by willvedd** [comment]
@lundybernard @kanocarra thanks for being patient. We've addressed this in our [7.5.0](https://github.com/auth0/auth0-deploy-cli/releases/tag/v7.5.0) release, I encourage you folks to update and test out. 

The approach taken can be looked at more closely in #430 but in brief, we've just added excluded properties back onto the PATCH payload, which makes the bi-directional exclusion more explicit. We did consider the `__IGNORE__` syntax, but it would've added much more surface area to the problem when this only affects connections. So for the sake of simplicity and backwards compatibility, we decided against it for now.

I appreciate the patience and feedback on this issue; don't ever want to have the usage of this tool result in destructive changes. Going to close this out but please raise another issue if you're experiencing any issues with the new release.
